### ISSUE TITLE
Switch mosy users from abandoned `libelf` to `elfutils`

### ISSUE THREAD

**Issue by trofi** [NixOS/nixpkgs#271473]
### Describe the bug

`libelf` did not see an update since 2014: https://github.com/Distrotech/libelf. `elfutils` is a popular API compatible (?) `ELF` parser that most users expect.

I suggest switching most users (or maybe even all?) from `libelf` to `elfutils`. A few examples which should very likely use `elfutils`:

```
$ git grep -F ', libelf'
...
pkgs/development/libraries/mesa/default.nix:, libelf, libvdpau
pkgs/os-specific/linux/iproute/default.nix:, db, iptables, libelf, libmnl
pkgs/os-specific/linux/kernel/manual-config.nix:, libelf, cpio, elfutils, zstd, python3Minimal, zlib, pahole, kmod, ubootTools
```

### Priorities

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by Artturin** [comment]
Arch has the `libelf` package as an output of the `elfutils` package

https://archlinux.org/packages/core/x86_64/libelf/

https://gitlab.archlinux.org/archlinux/packaging/packages/elfutils/-/blob/main/PKGBUILD?ref_type=heads#L142

**Comment by Atemu** [comment]
If it's API-compatible, I'd be in favour of removing the old, putting an alias in its place and replacing the remaining usages of `libelf` with `elfutils`.

**Comment by philiptaron** [comment]
I'm taking a crack at this.

**Comment by Atemu** [comment]
WDYM by "require"? As a library at build time? As an executable at build time?

**Comment by trofi** [comment]
ius on "Exotic Nix Targets" pointed out that `elfutils` is marked broken on `pkgsStatic` (`elfutils` loads backends dynamically).

That might limit the scope of transition to packages that already assume dynamic linking, like `mesa`. Or there is a chance to make the dependency optional just for dynamic case.

**Comment by philiptaron** [comment]
> WDYM by "require"? As a library at build time? As an executable at build time?

I'm not _totally_ sure. It's in the `buildInputs` for `pkgs/development/tools/misc/elfutils/default.nix`, and when I try to make the switch to `elfutils` in some packages, like [mgba](https://github.com/philiptaron/nixpkgs/commit/8dc67cb3305a12587545ae86493b88b426f5b2ec), the feature detection doesn't fire until `zstd` is added to the consuming package.

The [configure.ac](https://sourceware.org/git/?p=elfutils.git;a=blob;f=configure.ac;h=af5b6bf77dd2bd17fb4bb0695e83e55c6ffbbef9;hb=HEAD) from the elfutils library is pretty involved. Here's the summary from the end of the file:

```
AC_MSG_NOTICE([
=====================================================================
        elfutils: ${PACKAGE_VERSION} (eu_version: ${eu_version})
=====================================================================

    Prefix                             : ${prefix}
    Program prefix ("eu-" recommended) : ${program_prefix}
    Source code location               : ${srcdir}
    Maintainer mode                    : ${enable_maintainer_mode}
    build arch                         : ${ac_cv_build}

    CFLAGS=${CFLAGS}
    CXXFLAGS=${CXXFLAGS}

  RECOMMENDED FEATURES (should all be yes)
    gzip support                       : ${with_zlib}
    bzip2 support                      : ${with_bzlib}
    lzma/xz support                    : ${with_lzma}
    zstd support                       : ${with_zstd}
    zstd compression support           : ${with_zstd_compress}
    libstdc++ demangle support         : ${enable_demangler}
    File textrel check                 : ${enable_textrelcheck}
    Symbol versioning                  : ${enable_symbol_versioning}

  NOT RECOMMENDED FEATURES (should all be no)
    Experimental thread safety         : ${use_locks}
    install elf.h                      : ${install_elfh}

  OTHER FEATURES
    Deterministic archives by default  : ${default_ar_deterministic}
    Native language support            : ${USE_NLS}
    Extra Valgrind annotations         : ${use_vg_annotations}
    libdebuginfod client support       : ${enable_libdebuginfod}
    Debuginfod server support          : ${enable_debuginfod}
    Default DEBUGINFOD_URLS            : ${default_debuginfod_urls}

  EXTRA TEST FEATURES (used with make check)
    have bunzip2 installed (required)  : ${HAVE_BUNZIP2}
    have zstd installed                : ${HAVE_ZSTD}
    C++11                              : ${HAVE_CXX11}
    debug branch prediction            : ${use_debugpred}
    gprof support                      : ${use_gprof}
    gcov support                       : ${use_gcov}
    run all tests under valgrind       : ${use_valgrind}
    gcc undefined behaviour sanitizer  : ${use_undefined}
    gcc address sanitizer              : ${use_address}
    clang memory sanitizer             : ${use_msan}
    use rpath in tests                 : ${tests_use_rpath}
    test biarch                        : ${utrace_cv_cc_biarch}
])
```

> ius on "Exotic Nix Targets" pointed out that elfutils is marked broken on pkgsStatic (elfutils loads backends dynamically).

[Here's the change that marked it as broken](https://github.com/NixOS/nixpkgs/pull/170115), from 2022. It references [a thread](https://lists.fedorahosted.org/pipermail/elfutils-devel/2014-November/004223.html) from almost a decade ago. I wonder if that's still the case, and whether the suggested change is possible.

I note that `configure.ac`, linked above, [appears to support BUILD_STATIC](https://sourceware.org/git/?p=elfutils.git;a=blob;f=configure.ac;hb=HEAD#l417).

**Comment by Atemu** [comment]
In that case we might need to add zstd to elfutils' propagatedNativeBuildInputs.

**Comment by philiptaron** [comment]
> propagatedNativeBuildInputs

I would think if it's also a runtime dependency (as buildInputs suggests that it is?) that it ought to go into `propagatedBuildInputs`? 

I use this [gist](https://gist.github.com/CMCDragonkai/45359ee894bc0c7f90d562c4841117b5) to understand this business. But I definitely don't have it on lockdown.

**Comment by Atemu** [comment]
It depends on what `zstd` is being used here; as a library, executable or both.

**Comment by Artturin** [comment]
`zstd` is only a make dependency on arch and it's not in the pc file therefore it's not necessary to propagate it

**Comment by Artturin** [comment]
Nevermind I was looking at libelf

the elfutils libelf.pc has zlib and libzstd in requires.private 

```
prefix=/nix/store/3f2sdbc2lsapj5i0qvdd31d0chycc2ya-elfutils-0.190
exec_prefix=${prefix}
libdir=/nix/store/3f2sdbc2lsapj5i0qvdd31d0chycc2ya-elfutils-0.190/lib
includedir=/nix/store/k1hyq2bx37083ay5q0ryf7lw90g9z1a7-elfutils-0.190-dev/include

Name: libelf
Description: elfutils libelf library to read and write ELF files
Version: 0.190
URL: http://elfutils.org/

Libs: -L${libdir} -lelf
Cflags: -I${includedir}

Requires.private: zlib libzstd
```

**Comment by lolbinarycat** [comment]
> ius on "Exotic Nix Targets" pointed out that `elfutils` is marked broken on `pkgsStatic` (`elfutils` loads backends dynamically).
> 
> That might limit the scope of transition to packages that already assume dynamic linking, like `mesa`. Or there is a chance to make the dependency optional just for dynamic case.

technically, shouldn't it be possible for a statically linked library to call dlopen?

**Comment by trofi** [comment]
On `glibc` it might work today, but it will change eventually: https://sourceware.org/glibc/wiki/FAQ "In glibc 2.27, released in 2018, support for statically linked programs that call dlopen was deprecated. The intent was to simplify support for statically linked programs and avoid the problematic use cases that users created by mixing the two linkage models. Users should either link statically or dynamically, picking the model that best meets their needs while avoiding a hybrid mix of the two."

Static `musl` (which is what `pkgsStatic` on `nixpkgs` uses) would always fail at runtime on `dlopen()` calls: https://git.musl-libc.org/cgit/musl/tree/src/ldso/dlopen.c

**Comment by theoparis** [comment]
It seems like elfutils doesn't want to be installed on an aarch64-darwin system, and it is required to build a linux kernel 😕 
Is this something that can be supported in upstream elfutils?

**Comment by Atemu** [comment]
I don't think building our kernel drv ever worked on darwin. That's for another issue I think.

**Comment by philiptaron** [comment]
> Is this something that needs to be supported in upstream elfutils? Or is this a nixpkgs specific issue?

Yes, it's something that would need upstream support. See for instance https://github.com/avrdudes/avrdude/issues/1653 for Homebrew maintainers struggling with the same thing.

**Comment by liqili** [comment]
In macos there is no way to add the libelf package and elfutils package doesn't support for macos as well. I always got the following error then I have to use docker again.
fatal error: 'elf.h' file not found
